toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -56,29 +56,65 @@ func main() {
 	kingpin.MustParse(app.Parse(os.Args[1:]))
 	logger.InitBestEffort(*logFile, *logLevel)
 
+	err := fetchPackages()
+	if err != nil {
+		logger.Log.Fatalf("Failed to fetch packages. Error: %s", err)
+	}
+}
+
+func fetchPackages() (err error) {
 	dependencyGraph, err := pkggraph.ReadDOTGraphFile(*inputGraph)
 	if err != nil {
-		logger.Log.Panicf("Failed to read graph to file. Error: %s", err)
+		err = fmt.Errorf("failed to read graph to file: %w", err)
+		return
 	}
 
 	toolchainPackages, err := schedulerutils.ReadReservedFilesList(*toolchainManifest)
 	if err != nil {
-		logger.Log.Fatalf("unable to read toolchain manifest file '%s': %s", *toolchainManifest, err)
+		err = fmt.Errorf("unable to read toolchain manifest file '%s': %w", *toolchainManifest, err)
+		return
 	}
 
-	if hasUnresolvedNodes(dependencyGraph) {
-		err = resolveGraphNodes(dependencyGraph, *inputSummaryFile, *outputSummaryFile, toolchainPackages, *disableUpstreamRepos, *stopOnFailure)
+	var cloner *rpmrepocloner.RpmRepoCloner
+	hasUnresolvedNodes := hasUnresolvedNodes(dependencyGraph)
+	if hasUnresolvedNodes {
+		// Create the worker environment
+		cloner, err = prepRpmCloner(*outDir, *disableUpstreamRepos)
 		if err != nil {
-			logger.Log.Panicf("Failed to resolve graph. Error: %s", err)
+			err = fmt.Errorf("failed to setup new cloner: %w", err)
+			return
+		}
+		defer cloner.Close()
+	}
+
+	if hasUnresolvedNodes {
+		err = resolveGraphNodes(dependencyGraph, *inputSummaryFile, *outputSummaryFile, toolchainPackages, cloner, *stopOnFailure)
+		if err != nil {
+			err = fmt.Errorf("failed to resolve graph: %w", err)
+			return
 		}
 	} else {
 		logger.Log.Info("No unresolved packages to cache")
 	}
 
+	// Write the graph to file (even if we are going to download delta RPMs, we want to save the graph with the resolved nodes first)
 	err = pkggraph.WriteDOTGraphFile(dependencyGraph, *outputGraph)
 	if err != nil {
-		logger.Log.Panicf("Failed to write cache graph to file. Error: %s", err)
+		err = fmt.Errorf("failed to write cache graph to file: %w", err)
+		return
 	}
+
+	// If we grabbed any RPMs, we need to convert them into a local repo
+	if hasUnresolvedNodes {
+		logger.Log.Info("Configuring downloaded RPMs as a local repository")
```

Comment:
>logger.Log.Info("Configuring downloaded RPMs as a local repository") [](http://example.com/codeflow?start=2&length=68)

I'd either add one more info log around the time we try to resolve the graph nodes or reword this one, so that the understand what caused any RPMs to be downloaded. Right now this particular info message will appear for some builds (where we find unresolved nodes) but will be missing for others (where there were no unresolved nodes), so it might be a bit confusing.